app/services/fact_importer.rb

```diff
@@ -55,52 +57,53 @@ def fact_name_class
   attr_reader :host, :facts
 
   def delete_removed_facts
-    to_delete = host.fact_values.eager_load(:fact_name).where("fact_names.type = '#{fact_name_class}' AND fact_names.name NOT IN (?)", facts.keys)
-    # N+1 DELETE SQL, but this would allow us to use callbacks (e.g. auditing) when deleting.
-    deleted = to_delete.destroy_all
-    @counters[:deleted] = deleted.size
+    ActiveSupport::Notifications.instrument "fact_importer_deleted.foreman", :host_id => host.id, :host_name => host.name, :facts => facts, :deleted => [] do |payload|
+      # deletes all facts using a single SQL query (with inner query)
+      payload[:count] = @counters[:deleted] = FactValue.joins(:fact_name).where(:host => host, 'fact_names.type' => fact_name_class).where.not('fact_names.name' => facts.keys).delete_all
+    end
+  end
 
-    @db_facts = nil
-    logger.debug("Merging facts for '#{host}': deleted #{counters[:deleted]} facts")
+  def add_new_fact(fact_names, name)
+    method = host.new_record? ? :build : :create!
+    fact_name = create_fact_name(fact_names, name, facts[name])
+    host.fact_values.send(method, :value => facts[name], :fact_name => fact_name)
+  rescue => e
+    logger.error("Fact #{name} could not be imported because of #{e.message}")
+    @error = e
   end
 
   def add_new_facts
-    facts_to_create = facts.keys - db_facts.keys
-    # if the host does not exists yet, we don't have an host_id to use the fact_values table.
-    if facts_to_create.present?
-      method = host.new_record? ? :build : :create!
-      # :type is needed because custom facts usually inherits from FactName so they would be included in the list
-      fact_names = fact_name_class.where(:type => fact_name_class).index_by(&:name)
-      facts_to_create.each do |name|
-        begin
-          fact_name = create_fact_name(fact_names, name, facts[name])
-          host.fact_values.send(method, :value => facts[name], :fact_name => fact_name)
-        rescue => e
-          logger.error("Fact #{name} could not be imported because of #{e.message}")
-          @error = true
-        end
+    ActiveSupport::Notifications.instrument "fact_importer_added.foreman", :host_id => host.id, :host_name => host.name, :facts => facts do |payload|
+      facts_to_create = facts.keys - db_facts.pluck('fact_names.name')
+      # if the host does not exists yet, we don't have an host_id to use the fact_values table.
+      if facts_to_create.present?
+        # :type is needed because custom facts usually inherits from FactName so they would be included in the list
+        fact_names = fact_name_class.where(:type => fact_name_class).index_by(&:name)
```

Comment:
Would it make sense to also replace the `index_by(&:name)` here with `pluck(:name)`? I don't see any use for this hash other then for deciding which names need to be created.